### PR TITLE
kubectl: remove unnecessary TODO annotations

### PR DESCRIPTION
**Pull request by wujunwei**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
remove unnecessary TODO annotations
#### Does this PR introduce a user-facing change?

```docs

```

### CODE REVIEW

**Comment by wujunwei** [comment]
@dougsland @soltysh PTAL

File: staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go [hunk 1]
```diff
@@ -252,9 +252,11 @@ func (o *CopyOptions) Run() error {
 	}
 
 	if len(srcSpec.PodName) != 0 {
+		cmdutil.CheckErr(o.checkContainerCommandExist("tar", srcSpec))
```
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:255]: Is there a reason you're forcibly exiting (which is what CheckErr does) instead of returning the error?

Nit: I would slightly prefer these checks to be done from within copy functions, since the fact that those use tar is an implementation detail of each.

File: staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go [hunk 2]
```diff
@@ -304,7 +332,6 @@ func (o *CopyOptions) copyToPod(src, dest fileSpec, options *exec.ExecOptions) e
 	}(srcFile, destFile, writer)
 	var cmdArr []string
 
-	// TODO: Improve error messages by first testing if 'tar' is present in the container?
```
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:307]: There is a similar TODO at L376/403 below that should also be removed.

File: staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go [hunk 3]
```diff
@@ -283,6 +285,32 @@ func (o *CopyOptions) checkDestinationIsDir(dest fileSpec) error {
...
+			Namespace: dest.PodNamespace,
+			PodName:   dest.PodName,
+		},
+
+		Command:  []string{command},
+		Executor: &exec.DefaultRemoteExecutor{},
+	}
+
+	_ = o.execute(options)
+	if strings.Contains(outBuf.String(), fmt.Sprintf("\"%s\": executable file not found in $PATH", command)) {
```
Comment (KnVerey) [inline staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:308]: I'm concerned about portability with this error string matching approach. Can we check the error code instead? As @apelisse [pointed out](https://github.com/kubernetes/kubernetes/pull/79075#issuecomment-505659234) on a previous attempt at the same fix: 

> error code 127 usually means that the command was not found: https://www.tldp.org/LDP/abs/html/exitcodes.html

It might be good to cover code 126 too (found but not executable)

**Review by KnVerey — COMMENTED** [review]
cc @soltysh for review of any potential security concerns since this PR touches `kubectl cp`

File: staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go (hunk 3 above)
Comment (wujunwei) [reply staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:308]: I did it with checking error code `126`  at first. but I'm not sure if it returns the same code every time。 I thought it should be `127` ，but got `126` every time。

If you ensure that checking `126`,`127` is enougth,I will do that 

BTW，I will do it like below
```go
import utilexec "k8s.io/client-go/util/exec"


	err := o.execute(options)
	if err !=nil {
		if ce, ok:=err.(utilexec.CodeExitError);ok && ce.Code == 126{
			return  fmt.Errorf("command '%s' which kubectl need can not be found in the chosen container", command)
		}
	}
```

File: staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go (hunk 1 above)
Comment (wujunwei) [reply staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:255]: > Is there a reason you're forcibly exiting (which is what CheckErr does) instead of returning the error?
> 
> Nit: I would slightly prefer these checks to be done from within copy functions, since the fact that those use tar is an implementation detail of each.

I agree with you，I will move these checks within copy function。

**Comment by wujunwei** [comment]
pinging @KnVerey PTAL

**Review by KnVerey — COMMENTED** [review]
I like the change to the error-code-based approach, but I pulled this PR to try it out and it doesn't seem to work as expected. It seems like the error returned may not be a `utilexec.CodeExitError`, so the new condition isn't being hit. This points out something more important: this PR needs tests! Unless I'm missing something, the existing test coverage for `kubectl cp` in general seems to be quite sparse. In fact, I'm not seeing a single integration test in `test/cmd/`. Per the [Increasing the Reliability Bar](https://github.com/kubernetes/enhancements/pull/3139) initiative and [Open Letter](https://groups.google.com/a/kubernetes.io/g/dev/c/6F3h0Z1QzVg), I believe this command is a case where we need foundational test improvements to accompany functional changes.

**Comment by wujunwei** [comment]
> I like the change to the error-code-based approach, but I pulled this PR to try it out and it doesn't seem to work as expected. It seems like the error returned may not be a `utilexec.CodeExitError`, so the new condition isn't being hit. This points out something more important: this PR needs tests! Unless I'm missing something, the existing test coverage for `kubectl cp` in general seems to be quite sparse. In fact, I'm not seeing a single integration test in `test/cmd/`. Per the [Increasing the Reliability Bar](https://github.com/kubernetes/enhancements/pull/3139) initiative and [Open Letter](https://groups.google.com/a/kubernetes.io/g/dev/c/6F3h0Z1QzVg), I believe this command is a case where we need foundational test improvements to accompany functional changes.

I have try my PR locally , it works well,and I print the error return by o.execute(options).it's `exec.CodeExitError{Err:(*errors.errorString)(0xc0007525d0), Code:126}`  

Can you provide the steps can reproduct it?

BTW,I am glad to add some foundational tests about it ,but I don't know how to test about the action which need to interact with the containers, could you provide some examples or guides about it ?

**Comment by wujunwei** [comment]
> If I increase logging to `-v=6`, then I notice that the error is coming from the POST to actually attempt the copy operation, i.e. `checkContainerCommandExist` didn't stop it. If I change L313 to `return err` and build kubectl again, then I get the exact same error message, but from the `checkContainerCommandExist` request this time.

What 's the CRI you are using? containerd or something like cri-o? mine is containerd.  I have try the case you offered, it works as expected.
```
.\kubectl  cp go.mod  cptest:/one -v6
I1111 11:25:36.605038   21620 loader.go:373] Config loaded from file:  /home/adam/.kube/config
I1111 11:25:36.805061   21620 round_trippers.go:553] GET https://kubernetes.docker.internal:6443/api/v1/namespaces/default/pods/cptest 200 OK in 138 milliseconds
I1111 11:25:36.817676   21620 podcmd.go:88] Defaulting container name to cptest
I1111 11:25:37.099047   21620 round_trippers.go:553] POST https://kubernetes.docker.internal:6443/api/v1/namespaces/default/pods/cptest/exec?command=tar&container=cptest&stderr=true&std
out=true 101 Switching Protocols in 280 milliseconds
exec.CodeExitError{Err:(*errors.errorString)(0xc0006fe0f0), Code:126}
error: failed to copy file: 'tar': command which kubectl need can not be found in the chosen container
```
could you add the code `fmt.Printf("%#v", err)` to LIne 306 and try again :)

**Comment by KnVerey** [comment]
> What 's the CRI you are using? containerd or something like cri-o? mine is containerd. I have try the case you offered, it works as expected.

I'm on MacOS using Podman, which reports using `crun version 1.4.4`. I tried using Docker with `runc v1.0.1-0`, and it does work as expected in that case. This confirms it is runtime-related, and likely extremely difficult to test for even if CI did support running containers.

> could you add the code fmt.Printf("%#v", err) to LIne 306 and try again :)

Result:
```
&errors.errorString{s:"Internal error occurred: error executing command in container: failed to exec in container: failed to start exec \"b84fd4360b6dfea793310c1447aa14e21324f197c528c3426d55271fe35758e7\": OCI runtime exec failed: exec failed: unable to start container process: exec: \"tar\": executable file not found in $PATH: unknown"}error: Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "d6d0625b214fa5da782e0cef6727cc8ba9c6df40c540dffbaf918e6ac9434374": OCI runtime exec failed: exec failed: unable to start container process: exec: "tar": executable file not found in $PATH: unknown
```

**Comment by wujunwei** [comment]
> 
> Result:
> 
> ```
> &errors.errorString{s:"Internal error occurred: error executing command in container: failed to exec in container: failed to start exec \"b84fd4360b6dfea793310c1447aa14e21324f197c528c3426d55271fe35758e7\": OCI runtime exec failed: exec failed: unable to start container process: exec: \"tar\": executable file not found in $PATH: unknown"}error: Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "d6d0625b214fa5da782e0cef6727cc8ba9c6df40c540dffbaf918e6ac9434374": OCI runtime exec failed: exec failed: unable to start container process: exec: "tar": executable file not found in $PATH: unknown
> ```

I see, I think this error message also can let users understand why  the command failed.

**Comment by KnVerey** [comment]
> I see, I think this error message also can let users understand why the command failed.

This is the message that already gets printed today. If it is sufficient, do we need this PR or should we instead only remove the TODO?

**Comment by wujunwei** [comment]
> `&errors.errorString{s:"Internal error occurred: error executing command in container: failed to exec in container: failed to start exec \"b84fd4360b6dfea793310c1447aa14e21324f197c528c3426d55271fe35758e7\": OCI runtime exec failed: exec failed: unable to start container process: exec: \"tar\": executable file not found in $PATH: unknown"}error: Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "d6d0625b214fa5da782e0cef6727cc8ba9c6df40c540dffbaf918e6ac9434374": OCI runtime exec failed: exec failed: unable to start container process: exec: "tar": executable file not found in $PATH: unknown`

Have you tested  `kubectl cp  cptest:/bin   ./` in your case ? it will only show `command terminated with exit code 126`.  

so we do need this PR

**Comment by wujunwei** [comment]
Have you tested  `kubectl cp  cptest:/bin   ./` in your case ? it will only show `command terminated with exit code 126`.  

so we do need this PR .

**Comment by wujunwei** [comment]
> In all cases, the message mentions tar being the problem. Are you able to produce an error message that does not?

I tried using `containerd` + `runc`  I get this message:
```
command terminated with exit code 126
```
It cann't provide any available infomation, That's why I submitted this PR.

**Comment by KnVerey** [comment]
As promised, I brought up this PR in particular + cp testing in general in today's SIG CLI meeting. You can watch the recording here: https://youtu.be/ByAor52goTA. The feedback from my fellow leads is that although an e2e test suite that does run containers [exists and has some coverage of cp](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubectl/kubectl.go), we cannot accept changes that only work with or exist only to accommodate specific container runtimes. A concern regarding Windows in particular was also raised, and it was noted that `kubectl cp` is effectively [frozen at a very small feature set](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/2381-future-of-kubectl-cp) primarily for security reasons. In the long term, we would like copy exposed by the CRI itself, such that kubectl could be implemented without using tar at all. There seems to be a lot of support for that approach, but nobody available to work on it so far: https://github.com/kubernetes/kubernetes/issues/58512#issuecomment-780793551.

In short, we cannot accept the change in this PR, and we should remove the TODO from the code to stop signalling to contributors that we would. Would you be willing to update the PR to do only that? I'm very sorry that this is not the result you wanted, and that I did not have this information up front.

**Comment by wujunwei** [comment]
Got it. thanks for replying, maybe I should look into some CRI how to improve it.

> Would you be willing to update the PR to do only that?

I have update my PR , only remove TODO.

**Comment by KnVerey** [comment]
Thank you for understanding! Can you please also update your PR title and description (in particular removing the release note)?

**Comment by wujunwei** [comment]
> Thank you for understanding! Can you please also update your PR title and description (in particular removing the release note)?

Done :) !

**Comment by wujunwei** [comment]
It seems that it need  `v1.26 milestones`

**Comment by KnVerey** [comment]
We're in code freeze now, so this will wait to merge until the thaw on Dec 6, at which point it should happen automatically. Only critical bug fixes are eligible for the 1.26 milestone at this time.
